packages/arb-validator/checkpointing/indexedCheckpointer.go

```diff
@@ -0,0 +1,477 @@
+	key := append([]byte{1}, height.AsInt().Bytes()...)
+	idBufs := []*common.BlockIdBuf{}
+	for _, id := range ids {
+		idBufs = append(idBufs, id.MarshalToBuf())
+	}
+	idBufList := &BlockIdBufList{
+		Bufs: idBufs,
+	}
+	valBytes, err := proto.Marshal(idBufList)
+	if err != nil {
+		return err
+	}
+	ok := cp.db.SaveData(key, valBytes)
+	if !ok {
+		return errors.New("db write error in checkpointer.setIdsAtHeight")
+	}
+	return nil
+}
+
+func (cp *IndexedCheckpointer) deleteIdsAtHeight(height *common.TimeBlocks) error {
+	key := append([]byte{1}, height.AsInt().Bytes()...)
+	if ok := cp.db.DeleteData(key); !ok {
+		return errors.New("Checkpointer failed to delete idsAtHeight")
+	}
+	return nil
+}
+
+func (cp *IndexedCheckpointer) recordIdAsCheckpointed(newId *common.BlockId) error {
+	ids, err := cp.getIdsAtHeight(newId.Height)
+	if err != nil {
+		return err
+	}
+	ids = append(ids, newId)
+	return cp.setIdsAtHeight(newId.Height, ids)
+}
+
+func (cp *IndexedCheckpointer) GetInitialMachine() (machine.Machine, error) {
+	cp.Lock()
+	defer cp.Unlock()
+
+	return cp.db.GetInitialMachine()
+}
+
+type writableCheckpoint struct {
+	blockId  *common.BlockId
+	contents []byte
+	ckpCtx   CheckpointContext
+}
+
+func (cp *IndexedCheckpointer) AsyncSaveCheckpoint(
+	blockId *common.BlockId,
+	contents []byte,
+	cpCtx CheckpointContext,
+	closeWhenDone chan struct{},
+) {
+	cp.Lock()
+	defer cp.Unlock()
+
+	if closeWhenDone != nil {
+		cp.chansToClose = append(cp.chansToClose, closeWhenDone)
+	}
+	cp.nextCheckpointToWrite = &writableCheckpoint{
+		blockId:  blockId,
+		contents: contents,
+		ckpCtx:   cpCtx,
+	}
+}
+
+func (_ *IndexedCheckpointer) makeContentsKey(id *common.BlockId) []byte {
+	bidBuf := id.MarshalToBuf()
+	bytesBuf, err := proto.Marshal(bidBuf)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return append([]byte{2}, bytesBuf...)
+}
+
+func (cp *IndexedCheckpointer) RestoreLatestState(
+	ctx context.Context,
+	clnt arbbridge.ArbClient,
+	_ common.Address,
+	_ bool,
+	callback func([]byte, RestoreContext),
+) error {
+	cp.Lock()
+	defer cp.Unlock()
+
+	heightBounds, err := cp.getHeightBounds()
+	if err != nil {
+		return err
+	}
+	if heightBounds == nil {
+		return errors.New("Cannot restore because no checkpoint exists")
+	}
+	for height := heightBounds.hi; height.Cmp(heightBounds.lo) >= 0; height = common.NewTimeBlocks(new(big.Int).Sub(height.AsInt(), big.NewInt(1))) {
+		ids, err := cp.getIdsAtHeight(height)
+		if err != nil {
+			return err
+		}
+		onchainId, err := clnt.BlockIdForHeight(ctx, height)
+		if err != nil {
+			return err
+		}
+		for _, id := range ids {
+			if id.Equals(onchainId) {
+				key := cp.makeContentsKey(id)
+				val := cp.db.GetData(key)
+				ckpWithMan := &CheckpointWithManifest{}
+				if err := proto.Unmarshal(val, ckpWithMan); err != nil {
+					return err
+				}
+				callback(ckpWithMan.Contents, cp.newRestoreContextLocked())
+				return nil
+			}
+		}
+	}
+	callback(nil, nil)
```

Comment:
Does the callback know what to do when nil is passed as both params? This is fairly nonstandard looking. What does it mean if the code reaches this point?

Reply:
Replaced this with a log.Fatal, because this should never happen.